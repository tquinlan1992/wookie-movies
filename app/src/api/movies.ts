import axios from "axios";

// const sampleMovie = {
//     backdrop: "https://wookie.codesubmit.io/static/backdrops/d6822b7b-48bb-4b78-ad5e-9ba04c517ec8.jpg"
// cast: ['Christian Bale', 'Heath Ledger', 'Aaron Eckhart']
// classification: "13+"
// director: "Christopher Nolan"
// genres: ['Action', 'Crime', 'Drama']
// id: "d6822b7b-48bb-4b78-ad5e-9ba04c517ec8"
// imdb_rating: 9
// length: "2h 32min"
// overview: "Batman raises the stakes in his war on crime. With the help of Lt. Jim Gordon and District Attorney Harvey Dent, Batman sets out to dismantle the remaining criminal organizations that plague the streets. The partnership proves to be effective, but they soon find themselves prey to a reign of chaos unleashed by a rising criminal mastermind known to the terrified citizens of Gotham as the Joker."
// poster: "https://wookie.codesubmit.io/static/posters/d6822b7b-48bb-4b78-ad5e-9ba04c517ec8.jpg"
// released_on: "2008-07-16T00:00:00"
// slug: "the-dark-knight-2008"
//
// }

export interface Movie {
  backdrop: string;
  cast: string[];
  classification: string;
  director: string;
  genres: string[];
  id: string;
  imdb_rating: 9;
  length: string;
  overview: string;
  poster: string;
  released_on: string;
  slug: string;
  title: string;
}

export const getMovies = async (search: string) => {
  const moviesResponse = await axios.get<{ movies: Movie[] }>(
    "https://wookie.codesubmit.io/movies",
    {
      headers: { Authorization: "Bearer Wookie2021" },
      params: { q: search },
    }
  );
  return moviesResponse.data.movies;
};
