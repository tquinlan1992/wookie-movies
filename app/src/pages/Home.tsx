import React, { useEffect, useState } from "react";
import styled from "styled-components";
import { getMovies, Movie } from "../api/movies";
import MoviesList from "../components/MoviesList";

const PageWrapper = styled.div`
  padding: 15px;
`;

const Home: React.FC<{ search: string }> = ({ search }) => {
  const [movies, setMovies] = useState<Movie[]>([]);
  useEffect(() => {
    const loadMovies = async () => {
      setMovies(await getMovies(search));
    };
    loadMovies();
  }, [search]);
  return (
    <PageWrapper>
      <MoviesList movies={movies} />
    </PageWrapper>
  );
};

export default Home;
