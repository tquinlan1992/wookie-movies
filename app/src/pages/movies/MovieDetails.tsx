import _ from "lodash";
import { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import styled from "styled-components";
import { getMovies, Movie } from "../../api/movies";
import PageWrapper from "../../components/PageWrapper";

const Column1 = styled.div`
  width: 400px;
  margin: auto;
`;

const Column2 = styled.div`
  flex: 0 1 auto;
  flex-wrap: wrap;
`;

const MovieDetailsColumns = styled.div`
  display: flex;
  width: 100%;
  @media (max-width: 800px) {
    flex-direction: column;
  }
`;

const Poster = styled.img`
  width: auto;
  height: auto;
  max-height: 500px;
  display: block;
  margin: auto;
  padding-right: 20px;
  @media (max-width: 800px) {
    padding-right: 0px;
    padding-bottom: 20px;
  }
`;

const Title = styled.div`
  padding-bottom: 30px;
  @media (max-width: 800px) {
    text-align: center;
  }
`;

const useMovieData = () => {
  const { id } = useParams<{ id: string }>();
  const [movieData, setMovieData] = useState<Movie>();
  useEffect(() => {
    const retriveMovie = async () => {
      const result = await getMovies("");
      const movieDataFromResult = result.find((movie) => movie.id === id);
      setMovieData(movieDataFromResult);
    };
    retriveMovie();
  }, [id]);
  return movieData;
};

const MovieDetails: React.FC = () => {
  const movieData = useMovieData();

  if (!movieData) {
    return null;
  } else {
    return (
      <PageWrapper>
        <MovieDetailsColumns>
          <Column1>
            <Poster src={movieData.poster} alt="Poster" />
          </Column1>
          <Column2>
            <Title>
              MovieDetails: {movieData.title} ({movieData.classification})
            </Title>
            <div>
              <span>
                {new Date(movieData.released_on).getFullYear()} |{" "}
                {movieData.length} |{" "}
                {_.isArray(movieData.director)
                  ? movieData.director.join(", ")
                  : movieData.director}
              </span>
            </div>
            <div>
              <span>Cast: {movieData.cast.join(", ")}</span>
            </div>
            <div>
              <p>Movie Description: {movieData.overview}</p>
            </div>
          </Column2>
        </MovieDetailsColumns>
      </PageWrapper>
    );
  }
};

export default MovieDetails;
