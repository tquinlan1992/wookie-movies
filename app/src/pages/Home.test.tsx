import { configure, mount } from "enzyme";
import Home from "./Home";
import * as moviesApi from "../api/movies";
import toJson from "enzyme-to-json";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { act } from "react-dom/test-utils";
import { Movie } from "../api/movies";

configure({ adapter: new Adapter() });

const waitForComponentToPaint = async (wrapper: any) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();
  });
};

function partialAs<Type>(input: Partial<Type>) {
  return input as Type;
}

test("List movie titles", async () => {
  const getMoviesSpy = jest.spyOn(moviesApi, "getMovies").mockReturnValue(
    Promise.resolve([
      partialAs<Movie>({
        title: "movie1",
        id: "movie1Id",
        genres: ["genre1", "genre2"],
      }),
      partialAs<Movie>({
        title: "movie2",
        id: "movie2Id",
        genres: ["genre3", "genre4"],
      }),
      partialAs<Movie>({
        title: "movie3",
        id: "movie3Id",
        genres: ["genre4", "genre5"],
      }),
    ])
  );
  const wrapper = mount(<Home />);
  await waitForComponentToPaint(wrapper);
  expect(toJson(wrapper)).toMatchSnapshot();
  expect(getMoviesSpy).toHaveBeenCalled();
});
