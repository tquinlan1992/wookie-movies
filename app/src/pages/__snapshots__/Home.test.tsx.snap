// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List movie titles 1`] = `
<Home>
  <styled.div>
    <div
      className="sc-bdfBQB eAxuQk"
    >
      <ListMovies
        movies={
          Array [
            Object {
              "genres": Array [
                "genre1",
                "genre2",
              ],
              "id": "movie1Id",
              "title": "movie1",
            },
            Object {
              "genres": Array [
                "genre3",
                "genre4",
              ],
              "id": "movie2Id",
              "title": "movie2",
            },
            Object {
              "genres": Array [
                "genre4",
                "genre5",
              ],
              "id": "movie3Id",
              "title": "movie3",
            },
          ]
        }
      >
        <h3>
          genre1
        </h3>
        <span
          key="movie1Id"
        >
          movie1
        </span>
        <h3>
          genre2
        </h3>
        <span
          key="movie1Id"
        >
          movie1
        </span>
        <h3>
          genre3
        </h3>
        <span
          key="movie2Id"
        >
          movie2
        </span>
        <h3>
          genre4
        </h3>
        <span
          key="movie2Id"
        >
          movie2
        </span>
        <span
          key="movie3Id"
        >
          movie3
        </span>
        <h3>
          genre5
        </h3>
        <span
          key="movie3Id"
        >
          movie3
        </span>
      </ListMovies>
    </div>
  </styled.div>
</Home>
`;
